torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,195 @@
+import itertools
+from typing_extensions import Protocol
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class _LazyProtocol(Protocol):
+    """This is to avoid errors with mypy checks for 
+    The attributes in a mixin:
+    https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes
+    """
+    def _register_load_state_dict_pre_hook(self, hook):
+        ...
+
+    def register_forward_pre_hook(self, hook):
+        ...
+
+    def _lazy_load_hook(
+            self, state_dict, prefix, local_metadata, strict,
+            missing_keys, unexpected_keys, error_msgs):
+        ...
+
+    def _get_name(self):
+        ...
+
+    def _infer_parameters(self, module, input):
+        ...
+
+    @property
+    def _parameters(self):
+        ...
+
+    @property
+    def _buffers(self):
+        ...
+
+    @property
+    def _non_persistent_buffers_set(self):
+        ...
+
+    @property
+    def _load_hook(self):
+        ...
+
+    @property
+    def _initialize_hook(self):
+        ...
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
+    that you need to access the weights before the actual "real" ``forward`` pass.
+
+    >>> lazy_module = torch.nn.LazyLinear(10)
+    >>> lazy_module.forward(input)  # Dummy forward call to initialize the module parameters
```

Comment:
Let's use language like, "dry run initializes the lazy modules".

It would be helpful to show the before/after structure of a simple network here. For example, a MLP with two linear layers and a ReLU in-between them. Also, this should make clear that the input for the dry run must be of an appropriate size. It would also help if a transform, like putting the modules on CUDA, was applied to the network so users understand when to perform transforms like that, and then to show an optimizer being attached to the network, too.

Reply:
Added the example.